File: C++/Day07.cpp
#include "AH.h"

namespace Day07
{

	const std::string CARDS  = "AKQJT98765432";
	const std::string CARDSJ = "AKQT98765432J";

	std::pair<std::string, int> parseInput(std::string s)
	{
		const auto ss = AH::Split(s, ' ');
		const int bet = stoi(ss[1]);

		return std::make_pair(ss[0], bet);
	}

	int handRank(const std::string hand)
	{
		std::map<char, int> cardCount;
		for (auto c :  hand) {
			cardCount[c] += 1;
		}

		if (cardCount.size() == 1) { // all cards the same, must be Five of a kind
			return 6;
		}

		if (cardCount.size() == 5) { // all cards different, must be High card
			return 0;
		}

		int maxCard = 0;
		for (auto const& [key, val] : cardCount) {
			maxCard = std::max(maxCard, val);
		}

		if (maxCard == 4) { // most common card seen 4 times, must be Four of a Kind
			return 5;
		}

		if (maxCard == 3) { // either Full house or Three of a kind
			if (cardCount.size() == 2) {
				return 4; // Full House
			} else {
				return 3; // Three of kind
			}
		}

		if (maxCard == 2) { // either Two pair or One pair
			if (cardCount.size() == 3) {
				return 2; // Two pair
			} else {
				return 1; // One Pair
			}
		}

		// should never be hit, only here to stop compiler warning
		return 0;
	}

	// It is always optimal to promote jokers to the same card
	int handRankWithJokers(const std::string hand)
	{
		int best = 0;
		for (auto c : CARDS) {
			if (c == 'J') continue;
			auto s = hand;
			std::replace( s.begin(), s.end(), 'J', c);
			best = std::max(best, handRank(s));
		}

		return best;
	}

	bool handLT(std::string a, std::string b, const bool part2=false)
	{
		int a_score = 0, b_score = 0;
		if (!part2) {
			a_score = handRank(a);
			b_score = handRank(b);			
		} else {
			a_score = handRankWithJokers(a);
			b_score = handRankWithJokers(b);
		}

		if (a_score != b_score) {
			return a_score < b_score;
		}

		for (size_t i = 0; i < 5; ++i) {
			int a_i = 0, b_i = 0;
			if (!part2) {
				a_i = CARDS.find(a.at(i));
				b_i = CARDS.find(b.at(i));			
			} else {
				a_i = CARDSJ.find(a.at(i));
				b_i = CARDSJ.find(b.at(i));	
			}

			if (a_i != b_i) {
				return a_i > b_i;
			}
		}

		// should never be hit, only here to stop compiler warning
		return false;
	}

	int Run(const std::string& filename)
	{
		const std::vector<std::string> inputLines = AH::ReadTextFile(filename);
		std::vector<std::pair<std::string, int>>hands;
		for (auto & l : inputLines) {
			hands.push_back(parseInput(l));
		}

		int part1 = 0, part2 = 0;

		sort( hands.begin( ), hands.end( ), [ ]( const auto& lhs, const auto& rhs )
		{
		   return handLT(lhs.first, rhs.first);
		});

		int idx = 1;
		for (auto [h, i] : hands) {
			part1 += idx * i;
			++idx;
		}

		sort( hands.begin( ), hands.end( ), [ ]( const auto& lhs, const auto& rhs )
		{
		   return handLT(lhs.first, rhs.first, true);
		});

		idx = 1;
		for (auto [h, i] : hands) {
			part2 += idx * i;
			++idx;
		}

		AH::PrintSoln(7, part1, part2);

		return 0;
	}

}

File: Go/Day07.go
package main

import (
	"sort"
	"strconv"
	"strings"

	AH "./adventhelper"
)

var CARDS = "AKQJT98765432"
var CARDSJ = "AKQT98765432J"

type Hand struct {
	cards string
	bet   int
}

func parseInput(s string) Hand {
	ss := strings.Split(s, " ")
	bet, _ := strconv.Atoi(ss[1])

	return Hand{ss[0], bet}
}

func handRank(hand string) int {
	cardCount := make(map[rune]int)
	for _, r := range hand {
		cardCount[r] += 1
	}

	if len(cardCount) == 1 {
		return 6
	}

	if len(cardCount) == 5 {
		return 0
	}

	maxCard := 0
	for _, v := range cardCount {
		if maxCard < v {
			maxCard = v
		}
	}

	if maxCard == 4 {
		return 5
	}

	if maxCard == 3 {
		if len(cardCount) == 2 {
			return 4
		} else {
			return 3
		}
	}

	if maxCard == 2 {
		if len(cardCount) == 3 {
			return 2
		} else {
			return 1
		}
	}

	return 0
}

func handRankWithJokers(hand string) int {
	best := 0
	for _, c := range CARDS {
		if c == 'J' {
			continue
		}
		hand_copy := strings.Replace(hand, "J", string(c), 5)
		v := handRank(hand_copy)
		if best < v {
			best = v
		}
	}

	return best
}

func handLT(lhs string, rhs string, part2 bool) bool {
	score_lhs, score_rhs := 0, 0
	if !part2 {
		score_lhs = handRank(lhs)
		score_rhs = handRank(rhs)
	} else {
		score_lhs = handRankWithJokers(lhs)
		score_rhs = handRankWithJokers(rhs)
	}

	if score_lhs != score_rhs {
		return score_lhs < score_rhs
	}

	for i := 0; i < 5; i++ {
		l, r := 0, 0
		if !part2 {
			l = strings.Index(CARDS, string(lhs[i]))
			r = strings.Index(CARDS, string(rhs[i]))
		} else {
			l = strings.Index(CARDSJ, string(lhs[i]))
			r = strings.Index(CARDSJ, string(rhs[i]))
		}

		if l != r {
			return l > r
		}
	}

	return false
}

func main() {
	inputLines, _ := AH.ReadStrFile("../input/input07.txt")
	hands := []Hand{}
	for _, l := range inputLines {
		hands = append(hands, parseInput(l))
	}
	part1, part2 := 0, 0
	sort.Slice(hands, func(i, j int) bool {
		return handLT(hands[i].cards, hands[j].cards, false)
	})
	for i, h := range hands {
		part1 += (i + 1) * h.bet
	}

	sort.Slice(hands, func(i, j int) bool {
		return handLT(hands[i].cards, hands[j].cards, true)
	})
	for i, h := range hands {
		part2 += (i + 1) * h.bet
	}

	AH.PrintSoln(7, part1, part2)

	return
}
